allennlp/training/metrics.py

```diff
@@ -288,3 +290,196 @@ def reset(self):
+        """
+        if mask is None:
+            mask = torch.ones(gold_labels.size())
+        # If you actually passed in Variables here instead of Tensors, this will be a huge memory
+        # leak, because it will prevent garbage collection for the computation graph.  We'll ensure
+        # that we're using tensors here first.
+        if isinstance(predictions, Variable):
+            predictions = predictions.data.cpu()
+        if isinstance(gold_labels, Variable):
+            gold_labels = gold_labels.data.cpu()
+        if isinstance(mask, Variable):
+            mask = mask.data.cpu()
+
+        num_classes = predictions.size(-1)
+        if (gold_labels >= num_classes).any():
+            raise ConfigurationError("A gold label passed to SpanBasedF1Measure contains an "
+                                     "id >= {}, the number of classes.".format(num_classes))
+
+        sequence_lengths = get_lengths_from_binary_sequence_mask(mask)
+        argmax_predictions = torch.topk(predictions, 1, -1)[1].float().squeeze(-1)
+
+        # Iterate over timesteps in batch.
+        batch_size = gold_labels.size(0)
+        for i in range(batch_size):
+            sequence_prediction = argmax_predictions[i, :]
+            sequence_gold_label = gold_labels[i, :]
+            length = sequence_lengths[i]
+            prediction_spans = self._extract_spans(sequence_prediction[:length].tolist())
+            gold_spans = self._extract_spans(sequence_gold_label[:length].tolist())
+
+            for span in prediction_spans:
+                if span in gold_spans:
+                    self._true_positives[span[1]] += 1
+                    gold_spans.remove(span)
+                else:
+                    self._false_positives[span[1]] += 1
+            # These spans weren't predicted.
+            for span in gold_spans:
+                self._false_negatives[span[1]] += 1
+
+    def _extract_spans(self, tag_sequence: List[int]) -> Set[Tuple[Tuple[int, int], str]]:
+        """
+        Given an integer tag sequence corresponding to BIO tags, extracts spans.
+        Spans are inclusive and can be of zero length, representing a single word span.
+
+        Parameters
+        ----------
+        tag_sequence : List[int], required.
+            The integer class labels for a sequence.
+
+        Returns
+        -------
+        spans : Set[Tuple[Tuple[int, int], str]]
+            The typed, extracted spans from the sequence, in the format ((span_start, span_end), label).
+            Note that the label _does not_ contain any BIO tag prefixes.
+        """
+        spans = set()
+        span_start = 0
+        span_end = 0
+        active_conll_tag = None
+        for index, integer_tag in enumerate(tag_sequence):
+            # Actual BIO tag.
+            string_tag = self._label_vocabulary[integer_tag]
+            bio_tag = string_tag[0]
+            conll_tag = string_tag[2:]
+            if integer_tag in self._ignore_classes:
```

Comment:
Having now seen this logic, I think it'd be better to just take a set of `conll_tag` _strings_ as input to the constructor.  If the caller forgot to provide the index for "O" here, for example, this logic would be incorrect, but that's a really easy case to handle here.  If you do it this way, the caller can just say "ignore 'V' and 'ARG-LOC' spans", and it's easy, instead of having to query the dictionary for `B-V`, `I-V`, `B-ARG-LOC`, `I-ARG-LOC`, and `O`.  Looks like in your instantiation of the metric in the SRL model, you also forgot `I-V`.  Looking at the logic, I'm having a hard time telling if that will cause a bug or not.  Easier to reason about if the things you're ignoring are `conll_tags` instead of integer labels.

I think it would also make this piece of code easier to extend to the `C-` and other cases later, if we wanted to.